remove redundent "<>" for ty::Slice with reference type

lyming2007 opened the pull request:
this fix #103271

Commit:
remove redundent "<>" for ty::Slice with reference type
this fix #103271

fee1-dead reviewed (changes requested):
There might be a better way. AFAICT this diagnostic is erroneous because the span it covers does not contain the angle brackets, so it prints the suggestion on `[_]` instead of `<[_]>`. I believe this can be done pretty cheaply using [`span_to_source`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.span_to_source), skipping white space and enlarging the span if you see angle brackets from both sides.

lyming2007 commented:
> There might be a better way. AFAICT this diagnostic is erroneous because the span it covers does not contain the angle brackets, so it prints the suggestion on `[_]` instead of `<[_]>`. I believe this can be done pretty cheaply using [`span_to_source`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.span_to_source), skipping white space and enlarging the span if you see angle brackets from both sides.

pushed a new fix by seeing angle brackets from both sides.

fee1-dead commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1864,6 +1864,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        // <&[_]>::len doesn't need an extra "<>" between
+                        // but for Adt type like Vec::function()
+                        // we would suggest <[_]>::function();
+                        ty::Slice(_) if self.tcx.sess.source_map().span_wrapped_by_angle_bracket(ty.span)  => format!("{deref_ty}"),
                         _ => format!("<{deref_ty}>"),
```
I don't think it should just be `ty::Slice` as we recommend `<{deref_ty}>` to all types below. So we should stop recommending `<{deref_ty}>` no matter what `deref_ty` is if there are angle brackets already.

lyming2007 replied:
yes. You're right. We should not recommend <{deref_ty}> only for ty::Slice.
And the change was pushed.
Thank you for the suggestions.

fee1-dead reviewed (commented):
One last nit. Should be good to go after this.

fee1-dead commented:
Looks good to me, thanks for this PR!

@bors r+

bors commented:
:pushpin: Commit 868ccd5afefa4a6ea177eb68ec8f408e7d693a67 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead 868ccd5afefa4a6ea177eb68ec8f408e7d693a67 -->
<!-- homu: {"type":"Approved","sha":"868ccd5afefa4a6ea177eb68ec8f408e7d693a67","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Dylan-DPC commented:
failed in [rollup](https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882)

@bors r-

lyming2007 commented:
@fee1-dead should I do something for the merging conflict? This is the first time I encounter this type of conflict

Dylan-DPC commented:
What merging conflict? I don't see any conflict in your PR.

lyming2007 commented:
> failed in [rollup](https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882)



> failed in [rollup](https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882)

It says failed in https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882

Dylan-DPC commented:
yes it means your pr failed some tests when it was rolled up. You should see the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) (after a bit of scrolilng ;P). You need to update those tests and then we can merge this again

lyming2007 commented:
> yes it means your pr failed some tests when it was rolled up. You should see the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) (after a bit of scrolling ;P). You need to update those tests and then we can merge this again

I changed according to that error log. But the CI didn't pass:
```
---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
+	   = help: items from traits can only be used if the trait is in scope
+	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
+	   |
+	LL | use object::read::read_ref::ReadRef;
+	   |
10	help: the function `len` is implemented on `[_]`
11	   |
12	LL |     let length = <[_]>::len;
```

lyming2007 commented:
> yes it means your pr failed some tests when it was rolled up. You should see the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) (after a bit of scrolling ;P). You need to update those tests and then we can merge this again

I think my last code should fix the in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747). can you merge it again?

lyming2007 commented:
@Dylan-DPC any progress on it?

Dylan-DPC commented:
The CI doesn't seem to pass though? let me rerun it. Also i would wait for @fee1-dead 's confirmation whether it is fine to be merged again or not

lyming2007 commented:
> in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747)

I fixed it according to the error of this log [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747)
which seems conflict with the CI in PR https://github.com/rust-lang/rust/pull/103675

fee1-dead commented:
I have restarted the CI to see if this is fixed.

lyming2007 commented:
> I have restarted the CI to see if this is fixed.
the CI of this PR used to work.
It requires something like that:
```
---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
+	   = help: items from traits can only be used if the trait is in scope
+	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
+	   |
+	LL | use object::read::read_ref::ReadRef;
+	   |
10	help: the function `len` is implemented on `[_]`
11	   |
12	LL |     let length = <[_]>::len;
```
but the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) requires something like that:
```
failures:

---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
-	   = help: items from traits can only be used if the trait is in scope
-	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
-	   |
-	LL | use object::read::read_ref::ReadRef;
-	   |
15	help: the function `len` is implemented on `[_]`
16	   |
17	LL |     let length = <[_]>::len;


```
So they're conflicting with each other.  Can I remove the test case and you approve the PR? after the code is merged, I'll add the test case later. @fee1-dead

fee1-dead commented:
Could you try [rebasing](https://rustc-dev-guide.rust-lang.org/git.html#keeping-things-up-to-date) your PR on top of master?

fee1-dead commented:
Also to expand on my last point, the PR's CI must pass as if it fails then the test jobs will most likely fail after I approve. The failure of the rollup was `test-various`, which is only one of the multiple test jobs that we have. It could be spurious though and we could retry after CI passes.

lyming2007 commented:
> Also to expand on my last point, the PR's CI must pass as if it fails then the test jobs will most likely fail after I approve. The failure of the rollup was `test-various`, which is only one of the multiple test jobs that we have. It could be spurious though and we could retry after CI passes.

I've rebased the master and fixed the CI. Please advise. @fee1-dead

fee1-dead commented:
Let's see if it will work this time.

@bors r+

bors commented:
:pushpin: Commit c48ca7dc87c5efc5cf2797d3666790c057867ea4 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead c48ca7dc87c5efc5cf2797d3666790c057867ea4 -->
<!-- homu: {"type":"Approved","sha":"c48ca7dc87c5efc5cf2797d3666790c057867ea4","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez commented:
Failed in https://github.com/rust-lang/rust/pull/104151.

@bors r-

lyming2007 commented:
the test in https://github.com/rust-lang-ci/rust/actions/runs/3420197276/jobs/5694702753 failed as I mentioned before:
```
failures:

---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
-	   = help: items from traits can only be used if the trait is in scope
-	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
-	   |
-	LL | use object::read::read_ref::ReadRef;
-	   |
15	help: the function `len` is implemented on `[_]`
16	   |
17	LL |     let length = <[_]>::len;
```
but if I removed these code:
```
-	   = help: items from traits can only be used if the trait is in scope
-	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
-	   |
-	LL | use object::read::read_ref::ReadRef;
-	   |
```
This CI would fail at:
```
---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
+	   = help: items from traits can only be used if the trait is in scope
+	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
+	   |
+	LL | use object::read::read_ref::ReadRef;
+	   |
10	help: the function `len` is implemented on `[_]`
11	   |
12	LL |     let length = <[_]>::len;
```
@fee1-dead

fee1-dead commented:
I asked [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/CI.20and.20test-various.20contradicting.20results), and this is #26454. The solution here is to use another function name that does not have the suggestion.

lyming2007 commented:
@fee1-dead should be ok now. please approve it again. :)

fee1-dead commented on src/test/ui/type/issue-103271.rs:
```diff
@@ -0,0 +1,6 @@
+fn main() {
+    let length = <&[u32]>::len;
```
I think you need to change the `len` to another function that does not interfere with the trait from the sysroot.

lyming2007 replied:
Changed to `iter`

bors commented:
:pushpin: Commit 0b6934d6c6fed66e5e783b355de1f33ab3d51efb has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead 0b6934d6c6fed66e5e783b355de1f33ab3d51efb -->
<!-- homu: {"type":"Approved","sha":"0b6934d6c6fed66e5e783b355de1f33ab3d51efb","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
